File: web/src/ble/protocol.ts
/**
 * BLE GATT protocol constants — the single source of truth is docs/BLE_PROTOCOL.md.
 * Firmware, the web app, and the iOS app must all agree with that document.
 * All multi-byte fields are little-endian.
 */

/** Base 128-bit UUID; characteristic index is substituted into `f001000X-...`. */
export const SERVICE_UUID = 'f0010000-1a2b-4c3d-8e5f-000000000000';

export const CHAR_UUID = {
  control: 'f0010001-1a2b-4c3d-8e5f-000000000000',
  ecg: 'f0010002-1a2b-4c3d-8e5f-000000000000',
  bioz: 'f0010003-1a2b-4c3d-8e5f-000000000000',
  ppg: 'f0010004-1a2b-4c3d-8e5f-000000000000',
  metrics: 'f0010005-1a2b-4c3d-8e5f-000000000000',
} as const;

/** Standard Battery Service. */
export const BATTERY_SERVICE_UUID = 0x180f;
export const BATTERY_LEVEL_UUID = 0x2a19;

/** Control characteristic opcodes (first payload byte on write). */
export const Opcode = {
  START: 0x01,
  STOP: 0x02,
  SET_RATE: 0x03,
  SYNC_CLOCK: 0x04,
  GET_INFO: 0x05,
} as const;

/** Status notification message types (first payload byte). */
export const StatusMsg = {
  STATUS: 0x10,
  INFO: 0x11,
  CLOCK: 0x12,
} as const;

export type StreamKind = 'ecg' | 'bioz' | 'ppg';

/** Stream ids and START bitmask bits. */
export const StreamId: Record<StreamKind, number> = { ecg: 0, bioz: 1, ppg: 2 };
export const StreamBit: Record<StreamKind, number> = { ecg: 0x01, bioz: 0x02, ppg: 0x04 };

export const ALL_STREAMS_MASK = StreamBit.ecg | StreamBit.bioz | StreamBit.ppg;

/** Rate-code → Hz tables, per docs/BLE_PROTOCOL.md. */
export const RATE_CODES: Record<StreamKind, number[]> = {
  ecg: [128, 256, 512],
  bioz: [32, 64, 128],
  ppg: [50, 100, 200, 400],
};

export const DEFAULT_RATE_HZ: Record<StreamKind, number> = { ecg: 256, bioz: 64, ppg: 100 };

/** Default samples-per-packet used by the firmware / mock encoder (parsers derive K). */
export const DEFAULT_K: Record<StreamKind, number> = { ecg: 20, bioz: 16, ppg: 12 };

/** Error-flag bits in the STATUS message. */
export const ErrorFlag = {
  LEAD_OFF: 0x01,
  PPG_SATURATION: 0x02,
  BUFFER_OVERFLOW: 0x04,
} as const;

/** Invalid sentinels used in the metrics packet. */
export const Sentinel = {
  U16: 0xffff,
  I32_MIN: -2147483648,
  I16_MIN: -32768,
} as const;

/** Resolve a rate code to Hz, falling back to the stream default. */
export function rateFromCode(stream: StreamKind, code: number): number {
  return RATE_CODES[stream][code] ?? DEFAULT_RATE_HZ[stream];
}

/** Resolve a Hz value to its rate code (nearest match), for SET_RATE. */
export function codeFromRate(stream: StreamKind, hz: number): number {
  const codes = RATE_CODES[stream];
  let best = 0;
  let bestErr = Infinity;
  for (let i = 0; i < codes.length; i++) {
    const err = Math.abs(codes[i] - hz);
    if (err < bestErr) {
      bestErr = err;
      best = i;
    }
  }
  return best;
}

File: android/app/src/main/java/com/monivitals/ble/Protocol.kt
//
//  Protocol.kt
//  MoniVitals (Android)
//
//  BLE GATT protocol constants — the single source of truth is docs/BLE_PROTOCOL.md.
//  Firmware, web app, iOS app, and this Android app must all agree with that document.
//  All multi-byte fields are little-endian.
//
//  Ported from ios/MoniVitals/BLE/Protocol.swift (== web/src/ble/protocol.ts).
//

package com.monivitals.ble

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.UUID
import kotlin.math.abs

/** Stream kinds carried over their own characteristics. */
@Serializable
enum class StreamKind {
    @SerialName("ecg") ECG,
    @SerialName("bioz") BIOZ,
    @SerialName("ppg") PPG;

    companion object {
        val allCases: List<StreamKind> = listOf(ECG, BIOZ, PPG)
    }
}

object BLEProtocol {
    /** Base 128-bit service UUID. */
    val serviceUUID: UUID = UUID.fromString("f0010000-1a2b-4c3d-8e5f-000000000000")

    /** Characteristic UUIDs keyed by purpose. */
    object CharUUID {
        val control: UUID = UUID.fromString("f0010001-1a2b-4c3d-8e5f-000000000000")
        val ecg: UUID = UUID.fromString("f0010002-1a2b-4c3d-8e5f-000000000000")
        val bioz: UUID = UUID.fromString("f0010003-1a2b-4c3d-8e5f-000000000000")
        val ppg: UUID = UUID.fromString("f0010004-1a2b-4c3d-8e5f-000000000000")
        val metrics: UUID = UUID.fromString("f0010005-1a2b-4c3d-8e5f-000000000000")
    }

    /** Standard Battery Service (16-bit UUIDs expanded to their 128-bit form). */
    val batteryServiceUUID: UUID = UUID.fromString("0000180f-0000-1000-8000-00805f9b34fb")
    val batteryLevelUUID: UUID = UUID.fromString("00002a19-0000-1000-8000-00805f9b34fb")

    /** Client Characteristic Configuration Descriptor (for enabling notifications). */
    val cccdUUID: UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")

    /** Control characteristic opcodes (first payload byte on write). */
    object Opcode {
        const val START: Byte = 0x01
        const val STOP: Byte = 0x02
        const val SET_RATE: Byte = 0x03
        const val SYNC_CLOCK: Byte = 0x04
        const val GET_INFO: Byte = 0x05
    }

    /** Status notification message types (first payload byte). */
    object StatusMsg {
        const val STATUS: Int = 0x10
        const val INFO: Int = 0x11
        const val CLOCK: Int = 0x12
    }

    /** Stream ids (used in SET_RATE and conceptually in the START bitmask). */
    fun streamId(stream: StreamKind): Byte = when (stream) {
        StreamKind.ECG -> 0
        StreamKind.BIOZ -> 1
        StreamKind.PPG -> 2
    }

    /** START bitmask bit for a stream. */
    fun streamBit(stream: StreamKind): Int = when (stream) {
        StreamKind.ECG -> 0x01
        StreamKind.BIOZ -> 0x02
        StreamKind.PPG -> 0x04
    }

    const val allStreamsMask: Int = 0x01 or 0x02 or 0x04

    /** Rate-code → Hz tables, per docs/BLE_PROTOCOL.md. */
    fun rateCodes(stream: StreamKind): List<Int> = when (stream) {
        StreamKind.ECG -> listOf(128, 256, 512)
        StreamKind.BIOZ -> listOf(32, 64, 128)
        StreamKind.PPG -> listOf(50, 100, 200, 400)
    }

    /** Default sample rate (Hz) per stream. */
    fun defaultRateHz(stream: StreamKind): Int = when (stream) {
        StreamKind.ECG -> 256
        StreamKind.BIOZ -> 64
        StreamKind.PPG -> 100
    }

    /** Default samples-per-packet used by the firmware / mock encoder (parsers derive K). */
    fun defaultK(stream: StreamKind): Int = when (stream) {
        StreamKind.ECG -> 20
        StreamKind.BIOZ -> 16
        StreamKind.PPG -> 12
    }

    /** Error-flag bits in the STATUS message. */
    object ErrorFlag {
        const val leadOff: Int = 0x01
        const val ppgSaturation: Int = 0x02
        const val bufferOverflow: Int = 0x04
    }

    /** Invalid sentinels used in the metrics packet. */
    object Sentinel {
        const val u16: Int = 0xffff
        const val i32Min: Int = -2_147_483_648
        const val i16Min: Int = -32_768
    }

    /** Resolve a rate code to Hz, falling back to the stream default. */
    fun rate(stream: StreamKind, code: Int): Int {
        val codes = rateCodes(stream)
        return if (code in codes.indices) codes[code] else defaultRateHz(stream)
    }

    /** Resolve a Hz value to its rate code (nearest match), for SET_RATE. */
    fun code(stream: StreamKind, hz: Int): Int {
        val codes = rateCodes(stream)
        var best = 0
        var bestErr = Int.MAX_VALUE
        for (i in codes.indices) {
            val err = abs(codes[i] - hz)
            if (err < bestErr) {
                bestErr = err
                best = i
            }
        }
        return best
    }
}
